src/codegen/compile.ml

```diff
@@ -8626,21 +8193,222 @@ module Stabilization = struct
+        read_from_end env logical_size_offset I64Type ^^
+        StableMem.set_mem_size env ^^
+
+        (* clear size and version *)
+        clear_at_end env logical_size_offset I64Type ^^
+        clear_at_end env first_word_backup_offset I32Type ^^
+        clear_at_end env version_offset I32Type ^^
+
+        (* restore first word *)
+        compile_const_64 0L ^^ get_first_word ^^ StableMem.write_word32 env
+      end
+end
+
+module Persistence = struct
+  let load_stable_actor env = E.call_import env "rts" "load_stable_actor"
+    
+  let save_stable_actor env = E.call_import env "rts" "save_stable_actor"
+
+  let free_stable_actor env = E.call_import env "rts" "free_stable_actor"
+
+  let register_stable_type env actor_type =
+    let (candid_type_desc, type_offsets, type_indices) = Serialization.(type_desc env Persistence [actor_type]) in
+    let serialized_offsets = StaticBytes.(as_bytes [i32s (List.map Int32.of_int type_offsets)]) in
+    assert (type_indices = [0l]);
+    Blob.lit env candid_type_desc ^^
+    Blob.lit env serialized_offsets ^^
+    E.call_import env "rts" "register_stable_type"
+
+  let create_actor env actor_type get_field_value =
+    let (_, field_declarations) = Type.as_obj actor_type in
+    let field_initializers = List.map
+      (fun field -> (field.Type.lab, fun () -> (get_field_value field)))
+      field_declarations
+    in
+    Object.lit_raw env field_initializers
+
+  let recover_actor env actor_type =
+    let load_old_field env field =
+      if field.Type.typ = Type.(Opt Any) then
+        (* A stable variable may have been promoted to type `Any`: Therefore, drop its former content. *)
+        Opt.inject env (Tuple.compile_unit env)
+      else
+        (load_stable_actor env ^^ Object.load_idx_raw env field.Type.lab) in
+    let recover_field field = 
+      load_stable_actor env ^^
+      Object.contains_field env field.Type.lab ^^
+      (G.if1 I32Type
+        (load_old_field env field)
+        (Opt.null_lit env)
+      ) in
+    create_actor env actor_type recover_field ^^
+    free_stable_actor env
+
+  let save env actor_type =
+    save_stable_actor env ^^
+    NewStableMemory.backup env ^^
+    UpgradeStatistics.set_instructions env
+
+  let load env actor_type =
+    register_stable_type env actor_type ^^
+    load_stable_actor env ^^
+    G.i (Test (Wasm.Values.I32 I32Op.Eqz)) ^^
+    (G.if1 I32Type
+      begin
+        OldStabilization.old_destabilize env actor_type (NewStableMemory.upgrade_version env)
+      end
+      begin
+        recover_actor env actor_type ^^
+        NewStableMemory.restore env
+      end) ^^
+    StableMem.region_init env ^^
+    UpgradeStatistics.add_instructions env
+end
 
-  let store_static_roots env =
-    Arr.vanilla_lit env (E.get_static_roots env)
+module GCRoots = struct
+  let create_root_array env = Func.share_code0 Func.Always env "create_root_array" [I32Type] (fun env ->
+    E.(env.object_pool.frozen) := true;
```

Comment:
You may want to 'freeze' before the call to Func.shared_code0. I'm not sure the body is executed immediately but might might be delayed until the function code is generated.

Reply:
Thanks, I moved the check out of the function to be safe.